### DISCUSSION TITLE
estimate practice effects in non-decision time

### DISCUSSION THREAD

**Discussion by TinusUU** [mwshinn/PyDDM#111] [Ask for help]
Hi,

I want to model practice effect in non-decision time. I assume an exponential decay.  In addition, I expect this to be different for hard vs easy trials. I fit the model below over 500 trials, which is working fine, but very slow. 
Could this be done more efficient? Any suggestions? 

Best,
Martijn

```
m = pyddm.gddm(
    drift=lambda difficulty_value, v_hard, v_easy: v_hard * if difficulty_value == 0.04 else v_easy, 
    noise=1, 
    bound="b", 
    nondecision=lambda tr, tr_learn_hard, tr_learn_easy, tr_alpha_hard, tr_alpha_easy, difficulty_value, trial: (
        tr + tr_learn_hard * np.exp(-tr_alpha_hard * trial) if difficulty_value == 0.04 else
        tr + tr_learn_easy * np.exp(-tr_alpha_easy * trial)
        ),
    parameters={
        "v_easy": (0, 5), 
        "v_hard": (0, 5), 
        "tr": (0.1, 0.5),
        "tr_learn_hard": (0, 0.4),
        "tr_alpha_hard": (0, 0.2),
        "tr_learn_easy": (0, 0.4),
        "tr_alpha_easy": (0, 0.2),
        "b": (0.5, 2), 
    },
    choice_names=("Black", "White"),
    conditions=["difficulty_value", "trial"],
    T_dur=2, dx=.005, dt=.005
)
```

**Comment by mwshinn** [comment]
If you are using the latest version of PyDDM, this should have an analytical solution.  You can confirm this with m.has_analytical_solution().  You also have a typo on the second line ("* if").  But this should be pretty fast.

**Reply by TinusUU** [reply]
Ah yes: the typo was a residual from cleaning up the model for this post, which has an extra parameter (stim = 1 or -1) for drift-rate. 

Other than that, you are right that it's not thát slow, but much slower relative to the model with just tr.  
(fyi: I use a similar approach in a RLDDM, where adding these practice effects in Tr results in significant slower fits per participant).

However, I just tried a possible solution by binning trial-numbers and add a column with the average values of trial-numbers for each bin. Depending on bin-size, this helps a lot. Parameter estimates are in the ballpark of the values when using the full trial-range.
